File: aria-backend/src/aria/agents/rca-agent.ts
import { BedrockReasoner } from "../connectors/bedrock";
import { MongoRunbookConnector } from "../connectors/mongodb";
import { Neo4jConnector } from "../connectors/neo4j";
import { mockRcaResult } from "../mock-data";
import {
  AlertPayload,
  Hypothesis,
  InvestigationResult,
  RCAResult,
  Runbook,
  TriageResult,
} from "../types";

function buildFallbackHypotheses(
  alert: AlertPayload,
  investigation: InvestigationResult,
  runbooks: Runbook[],
): Hypothesis[] {
  const topError = investigation.datadog.topErrors[0];
  const topErrorMsg = topError?.message ?? "No error data available";
  const topErrorKind = topError?.errorKind;
  const primaryRunbook = runbooks[0];

  // Derive hypothesis title from actual error kind when available
  const primaryTitle = topErrorKind
    ? `${topErrorKind} in ${alert.service}`
    : `Elevated errors in ${alert.service}`;

  const primaryEvidence = [
    `Primary error: ${topErrorMsg}`,
    `Error rate ${alert.errorRatePct}% and p99 latency ${alert.p99LatencyMs}ms.`,
  ];

  if (investigation.datadog.metricTrend) {
    primaryEvidence.push(`Error trend: ${investigation.datadog.metricTrend.toUpperCase()}.`);
  }

  if (primaryRunbook) {
    primaryEvidence.push(`Historical runbook match: ${primaryRunbook.title}`);
  }

  return [
    {
      title: primaryTitle,
      probability: 0.85,
      evidence: primaryEvidence,
      remediation: primaryRunbook?.steps?.length
        ? primaryRunbook.steps.slice(0, 3)
        : [
            `Investigate recent deploys to ${alert.service} — rollback if correlated.`,
            "Check downstream service health for dependency failures.",
            "Scale replicas if load-driven.",
          ],
    },
    {
      title: "Recent deployment introduced regression",
      probability: 0.6,
      evidence: [
        "Error spike pattern is consistent with a bad deploy (sudden onset, constant error kind).",
        investigation.datadog.tracesSummary,
      ],
      remediation: [
        `Review recent ${alert.service} deployment history.`,
        "Rollback to last known good version if error rate persists > 5 minutes.",
      ],
    },
  ];
}

function pickRecommendedPlan(hypotheses: Hypothesis[], runbooks: Runbook[]): string[] {
  const dedupe = new Set<string>();
  const plan: string[] = [];

  for (const action of hypotheses[0]?.remediation ?? []) {
    if (!dedupe.has(action)) {
      dedupe.add(action);
      plan.push(action);
    }
  }

  for (const runbook of runbooks) {
    for (const step of runbook.steps) {
      if (!dedupe.has(step) && plan.length < 5) {
        dedupe.add(step);
        plan.push(step);
      }
    }
  }

  return plan;
}

export class RCAAgent {
  constructor(
    private readonly neo4j = new Neo4jConnector(),
    private readonly runbooks = new MongoRunbookConnector(),
    private readonly reasoner = new BedrockReasoner(),
  ) {}

  async run(
    alert: AlertPayload,
    _triage: TriageResult,
    investigation: InvestigationResult,
  ): Promise<RCAResult> {
    const [graph, runbooks] = await Promise.all([
      this.neo4j.fetchBlastRadius(alert.service),
      this.runbooks.fetchRunbooks(alert.service, alert.summary, 3),
    ]);

    const synthesized = await this.reasoner.synthesize({
      service: alert.service,
      summary: alert.summary,
      p99LatencyMs: alert.p99LatencyMs,
      errorRatePct: alert.errorRatePct,
      topErrors: investigation.datadog.topErrors,
      blastRadius: graph.impactedServices,
      runbooks,
      tracesSummary: investigation.datadog.tracesSummary,
      metricTrend: investigation.datadog.metricTrend,
      spanSummary: investigation.datadog.spanSummary,
      errorRateSeries: investigation.datadog.errorRateSeries,
      crossServiceLogs: investigation.datadog.crossServiceLogs,
    });

    if (synthesized) {
      const sortedHypotheses = [...synthesized.hypotheses].sort(
        (a, b) => b.probability - a.probability,
      );

      return {
        hypotheses: sortedHypotheses,
        blastRadius: graph.impactedServices,
        runbooks,
        recommendedPlan: synthesized.recommendedPlan,
        confidence: synthesized.confidence,
        narrative: synthesized.narrative,
      };
    }

    const fallback = mockRcaResult(alert.service);
    const hypotheses = buildFallbackHypotheses(alert, investigation, runbooks);

    return {
      ...fallback,
      hypotheses,
      blastRadius: graph.impactedServices.length
        ? graph.impactedServices
        : fallback.blastRadius,
      runbooks: runbooks.length ? runbooks : fallback.runbooks,
      recommendedPlan: pickRecommendedPlan(hypotheses, runbooks),
      confidence: hypotheses[0]?.probability ?? fallback.confidence,
      narrative:
        "Primary signal points to datastore saturation causing queue wait amplification across payment critical path.",
    };
  }
}

File: aria-backend/aria/agents/rca_agent.py
"""RCA Agent — Strands Agent with Neo4j (blast radius) and MongoDB (runbooks) tools."""
from __future__ import annotations

import asyncio
import json
import logging

from aria.config import config
from aria.connectors import mongodb as mongo_connector
from aria.connectors import neo4j as neo4j_connector
from aria.connectors.bedrock import synthesize_rca
from aria.mock_data import mock_runbooks
from aria.types import AlertPayload, Hypothesis, InvestigationResult, Runbook, TriageResult

logger = logging.getLogger(__name__)


def _extract_json(text: str) -> dict | None:
    start, end = text.find("{"), text.rfind("}")
    if start == -1 or end == -1:
        return None
    try:
        return json.loads(text[start : end + 1])
    except json.JSONDecodeError:
        return None


def _build_fallback(alert: AlertPayload, investigation: InvestigationResult, runbooks: list[Runbook], blast: list[str]) -> dict:
    top_error = investigation.datadog.topErrors[0].message if investigation.datadog.topErrors else "No top error"
    primary = runbooks[0] if runbooks else None
    hypotheses = [
        Hypothesis(
            title="Database connection pool saturation",
            probability=0.88,
            evidence=[
                f"Primary error: {top_error}",
                f"Error rate {alert.errorRatePct}% and p99 {alert.p99LatencyMs}ms consistent with saturation.",
                f"Matched runbook: {primary.title}" if primary else "No matching runbook found.",
            ],
            remediation=(primary.steps[:3] if primary and primary.steps else ["Increase DB pool limit.", "Add circuit breaker."]),
        ),
        Hypothesis(
            title="Downstream DB latency amplified by retry storm",
            probability=0.58,
            evidence=[
                "Retries and timeouts increase queue depth under contention.",
                "Trace summary shows DB wait dominates execution time.",
            ],
            remediation=["Throttle retries with jitter.", "Cap concurrent in-flight DB operations."],
        ),
    ]
    seen: set[str] = set()
    plan: list[str] = []
    for step in hypotheses[0].remediation:
        if step not in seen:
            seen.add(step)
            plan.append(step)
    for rb in runbooks:
        for step in rb.steps:
            if step not in seen and len(plan) < 5:
                seen.add(step)
                plan.append(step)
    return {
        "hypotheses": [h.model_dump() for h in hypotheses],
        "blastRadius": blast,
        "runbooks": [rb.model_dump() for rb in runbooks],
        "recommendedPlan": plan,
        "confidence": 0.88,
        "narrative": "Primary signal points to datastore saturation causing queue wait amplification across the payment critical path.",
    }


def _run_strands(alert: AlertPayload, triage: TriageResult, investigation: InvestigationResult) -> dict | None:
    try:
        from strands import Agent, tool
        from strands.models.bedrock import BedrockModel

        @tool
        def get_blast_radius(service: str) -> str:
            """Get the list of services impacted by a failure in the given service using the Neo4j dependency graph."""
            graph = neo4j_connector.fetch_blast_radius(service)
            return json.dumps({"impacted": graph.impactedServices, "upstream": graph.upstreamServices, "mode": graph.connectorMode})

        @tool
        def get_runbooks(service: str, query: str) -> str:
            """Retrieve historical incident runbooks matching the given service and incident description."""
            runbooks = mongo_connector.fetch_runbooks(service, query, 3)
            return json.dumps([rb.model_dump() for rb in runbooks])

        model = BedrockModel(model_id=config.bedrock_model_id, region_name=config.aws_region)
        agent = Agent(
            model=model,
            tools=[get_blast_radius, get_runbooks],
            system_prompt=(
                "You are ARIA's root-cause analyst. "
                "Call get_blast_radius and get_runbooks to gather evidence, "
                "then respond ONLY with JSON: "
                '{"narrative": str, "confidence": float, '
                '"hypotheses": [{"title": str, "probability": float, "evidence": [str], "remediation": [str]}], '
                '"recommendedPlan": [str]}.'
            ),
        )
        top_errors = [e.message for e in investigation.datadog.topErrors[:3]]
        prompt = (
            f"Service: {alert.service}. Severity: {triage.severity}. "
            f"Summary: {alert.summary}. "
            f"Top errors: {top_errors}. "
            "Call both tools then synthesize root cause."
        )
        result = agent(prompt)
        data = _extract_json(str(result))
        if data and "hypotheses" in data:
            return data
    except Exception as exc:
        logger.warning("Strands RCA failed (%s), falling back to direct synthesis.", exc)
    return None


class RCAAgent:
    async def run(self, alert: AlertPayload, triage: TriageResult, investigation: InvestigationResult) -> dict:
        # Always fetch graph + runbooks (needed for fallback too)
        graph, runbooks = await asyncio.gather(
            asyncio.to_thread(neo4j_connector.fetch_blast_radius, alert.service),
            asyncio.to_thread(mongo_connector.fetch_runbooks, alert.service, alert.summary, 3),
        )

        if config.connector_live("bedrock"):
            # Try Strands first
            result = await asyncio.to_thread(_run_strands, alert, triage, investigation)
            if result:
                result["blastRadius"] = graph.impactedServices
                result["runbooks"] = [rb.model_dump() for rb in runbooks]
                result["hypotheses"] = sorted(result.get("hypotheses", []), key=lambda h: -h.get("probability", 0))
                return result

            # Try direct Bedrock synthesis
            context = {
                "service": alert.service,
                "summary": alert.summary,
                "p99LatencyMs": alert.p99LatencyMs,
                "errorRatePct": alert.errorRatePct,
                "topErrors": [e.message for e in investigation.datadog.topErrors],
                "blastRadius": graph.impactedServices,
                "runbooks": [rb.model_dump() for rb in runbooks],
            }
            synthesized = await asyncio.to_thread(synthesize_rca, context)
            if synthesized and "hypotheses" in synthesized:
                synthesized["blastRadius"] = graph.impactedServices
                synthesized["runbooks"] = [rb.model_dump() for rb in runbooks]
                synthesized["hypotheses"] = sorted(synthesized["hypotheses"], key=lambda h: -h.get("probability", 0))
                return synthesized

        return _build_fallback(alert, investigation, runbooks, graph.impactedServices)
